### DISCUSSION TITLE
Where I suppose to put this function for  my restful api call ?

### DISCUSSION THREAD

**Discussion by chunji08** [behave/behave#913] [Q&A] [answered]
I have such a behave project, where include both UI and API level test cases. 
```
├── features
│   ├── api_tests
│   │   └── api_blueprints.feature
│   ├── blueprint
│   │   └── ui_blueprint.feature
│   ├── environment.cfg
│   ├── environment.py
│   ├── lib
│   │   └── pages
│   │       ├── base.py
│   │       ├── blueprint
│   │       └── utils.py
│   ├── requirements.txt
│   └── steps
│       ├── api_blueprint_part1.py
│       ├── api_blueprint_part2.py
│       └── ui_blueprint.py
└── README.md

```
As you can see, I am trying to use an existing UI test framework to add my RESTFUL API  test cases now, in which: 
1. **`features/api_tests/api_blueprints.feature`** defines all the API scenarios. 
2. **`features/steps/api_blueprint_part1.py`** and  **`features/steps/api_blueprint_part2.py`** contain all the **when/step/then** steps in python code. 

Now, all my api cases within the about 2 python files will eventually make the actual API calls by this function.  
```
 def execution_api_call(context, api_name, data_body="", action=None): 
        api_url = f'{context.base_url}{api_name}'
        token_setting = f"bearer {context.access_token}"
        headers = {"Authorization": token_setting}

        if action is None:
            with requests.session() as session:
                with session.get(api_url, headers=headers) as response:
                    context.response = response
       .....
```

in which: 
- context holds the base_url and access_token value, 
- api_name is unique per each call. 
- data_body depends on the individual api call. 
- action varies of: GET, POST, PUT or DELETE.
- and after the API execution, the response is saved back to context for future verification. 

Now my question is: where to put this function ? 
if I was to create a separate python file under features/step/.. to included it, it will not be visible to " api_blueprint_part1.py" or " api_blueprint_part2.py" when behave is being executed ? 
if I was to have it defined under features/lib/pages, I am not sure I should pass this "context" object within that call ? 

Comments ? 

Thanks, 

Jack

**Comment by jenisys** [comment, marked as answer]
Behave uses Python. Therefore, you have two options:

* Put shared functionality in a Python module (on the Python search path; PYTHONPATH environment variable)
* Put shared functionality in a Python module in the `steps/` directory

In both cases, you need to import the Python module first (in your step file(s)) before you can use/call its functions.

EXAMPLE:
* https://github.com/behave/behave.example/tree/master/features/steps (step_tutorial03.py uses collocated „blender.py“)

RELATED:
* Use a step library (normally collected to the „features/„ directory or installed)
